### PR TITLE
remember redirect_url when solving the 2FA challenge

### LINKED ISSUE #24850: Remember redirect_url when logging in with a second factor
While solving the 2FA challenge the redirect url is lost.

### Steps to reproduce
1. Enable a 2FA provider
2. Go to a ownCloud page that needs login
3. Get redirected to the login page (redirect url is encoded in the url)
4. Solve 2FA challenge

### Expected behaviour
The user should be redirected to the URL that was requested originally.

### Actual behaviour
The user is redirected to the files app.

**ownCloud version:**
9.1.0

ref #24559

### PR DESCRIPTION
**Pull request by ChristophWurst**
fixes #24850

@PVince81 @nickvergessen @rullzer review please

### CODE REVIEW

**Comment by ChristophWurst** [comment]
I created a simple test provider app: https://github.com/ChristophWurst/twofactor_test

File: core/Controller/TwoFactorChallengeController.php [hunk 1]
```diff
@@ -113,21 +118,28 @@ public function showChallenge($challengeProviderId) {
...
+				return new RedirectResponse($this->urlGenerator->getAbsoluteURL(urldecode($redirect_url)));
+			}
 			return new RedirectResponse($this->urlGenerator->linkToRoute('files.view.index'));
 		}
 
 		$this->session->set('two_factor_auth_error', true);
-		return new RedirectResponse($this->urlGenerator->linkToRoute('core.TwoFactorChallenge.showChallenge', ['challengeProviderId' => $provider->getId()]));
+		return new RedirectResponse($this->urlGenerator->linkToRoute('core.TwoFactorChallenge.showChallenge', [
+			'challengeProviderId' => $provider->getId(),
+			'redirect_url' => $redirect_url,
```
Comment (nickvergessen) [inline core/Controller/TwoFactorChallengeController.php]: needs to urlencode ?

**Comment by nickvergessen** [comment]
Works in general, but I'm loosing the `&fileid=` argument somewhere in the process :(

File: core/Controller/TwoFactorChallengeController.php (hunk 1 above)
Comment (ChristophWurst) [inline core/Controller/TwoFactorChallengeController.php]: is done here https://github.com/owncloud/core/blob/7b4459d28d40523c70ec05a733e158f2c14faac4/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php#L200

**Comment by nickvergessen** [comment]
:+1:

**Comment by ChristophWurst** [comment]
@PVince81 @rullzer second review? :-)

**Comment by PVince81** [comment]
:+1:

**Pull request merged**